presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java

```diff
@@ -384,13 +388,15 @@ private PlanRoot doAnalyzeQuery()
         stateMachine.setOutput(output);
 
         // fragment the plan
-        SubPlan fragmentedPlan = planFragmenter.createSubPlans(stateMachine.getSession(), plan, false, idAllocator, stateMachine.getWarningCollector());
+        // the variableAllocator is finally passed to SqlQueryScheduler for runtime cost-based optimizations
+        PlanVariableAllocator variableAllocator = new PlanVariableAllocator(plan.getTypes().allVariables());
+        SubPlan fragmentedPlan = planFragmenter.createSubPlans(stateMachine.getSession(), plan, false, idAllocator, variableAllocator, stateMachine.getWarningCollector());
 
         // record analysis time
         stateMachine.endAnalysis();
 
         boolean explainAnalyze = analysis.getStatement() instanceof Explain && ((Explain) analysis.getStatement()).isAnalyze();
-        return new PlanRoot(fragmentedPlan, !explainAnalyze, extractConnectors(analysis));
+        return new PlanRoot(fragmentedPlan, !explainAnalyze, extractConnectors(analysis), variableAllocator);
```

Comment:
This probably already raised discussions . But it looks weird to have `variableAllocator` in `PlanRoot`.

I see `PlanRoot#getVariableAllocator` is called when creating `SqlQueryScheduler`. Is it possible to create the `variableAllocator`  by using the plan? (i.e. `new PlanVariableAllocator(plan.getTypes().allVariables())`